gptqmodel/utils/importer.py

```diff
@@ -1,35 +1,73 @@
+from collections import OrderedDict
 from logging import getLogger
 
+from ..quantization import FORMAT
 from .backend import Backend
 
-
 logger = getLogger(__name__)
 
+from ..nn_modules.qlinear.qlinear_bitblas import QuantLinear as BitBLASQuantLinear
+from ..nn_modules.qlinear.qlinear_cuda import QuantLinear as CudaQuantLinear
+from ..nn_modules.qlinear.qlinear_cuda_old import QuantLinear as CudaOldQuantLinear
+from ..nn_modules.qlinear.qlinear_exllama import QuantLinear as ExllamaQuantLinear
+from ..nn_modules.qlinear.qlinear_exllamav2 import QuantLinear as ExllamaV2QuantLinear
+from ..nn_modules.qlinear.qlinear_marlin import QuantLinear as MarlinQuantLinear
+from ..nn_modules.qlinear.qlinear_tritonv2 import QuantLinear as TritonV2QuantLinear
+
+backend_dict = OrderedDict({
+    Backend.MARLIN: MarlinQuantLinear,
+    Backend.EXLLAMA_V2: ExllamaV2QuantLinear,
+    Backend.EXLLAMA: ExllamaQuantLinear,
+    Backend.TRITON: TritonV2QuantLinear,
+    Backend.CUDA_OLD: CudaOldQuantLinear,
+    Backend.CUDA: CudaQuantLinear,
+    Backend.BITBLAS: BitBLASQuantLinear,
+})
+
+format_dict = {
+    FORMAT.GPTQ: [Backend.EXLLAMA_V2, Backend.EXLLAMA, Backend.CUDA_OLD, Backend.CUDA],
+    FORMAT.GPTQ_V2: [Backend.EXLLAMA_V2, Backend.EXLLAMA, Backend.CUDA_OLD, Backend.CUDA],
+    FORMAT.MARLIN: [Backend.MARLIN],
+    FORMAT.BITBLAS: [Backend.BITBLAS],
+}
+
 
 # auto select the correct/optimal QuantLinear class
 def select_quant_linear(
-    bits: int,
-    group_size: int,
-    desc_act: bool,
-    sym: bool,
-    backend: Backend,
+        bits: int,
+        group_size: int,
+        desc_act: bool,
+        sym: bool,
+        backend: Backend,
+        format: str,
+        pack: bool = False,
 ):
-    # TODO handle AUTO
+    # Handle the case where backend is AUTO.
+    if backend == Backend.AUTO:
+        allow_backends = format_dict[format]
+        for k, v in backend_dict.items():
+            in_allow_backends = k in allow_backends if pack else True
+            validate = v.validate(bits, group_size, desc_act, sym, raise_error=True)
```

Comment:
I think you need to change raise_error=False so it doesn't crash here?

Reply:
updated.